ratings/ratings.py

```diff
@@ -0,0 +1,307 @@
+    Then, take the ten highest players with the highest uncertainty
+    For each of them, call them `p1`
+    Sort all the models by their distance from p1's rating and take the 20
+    nearest rated models. ('candidate_p2s')
+    Choose pairings, (p1, p2), randomly from this list.
+
+    'ratings' is a list of (model_num, rating, uncertainty) tuples
+
+    Returns a list of *model numbers*, not model ids.
+    """
+    db = sqlite3.connect("ratings.db")
+    data = db.execute("select model_winner, model_loser from wins").fetchall()
+    bucket_ids = [id[0] for id in db.execute(
+        "select id from models where bucket = ?", (fsdb.models_dir(),)).fetchall()]
+    bucket_ids.sort()
+    data = [d for d in data if d[0] in bucket_ids and d[1] in bucket_ids]
+
+    ratings = [(model_num_for(k), v[0], v[1]) for k, v in compute_ratings(data).items()]
+    ratings.sort()
+    ratings = ratings[100:]  # Filter off the first 100 models, which improve too fast.
+
+    ratings.sort(key=lambda r: r[2], reverse=True)
+
+    res = []
+    for p1 in ratings[:top_n]:
+        candidate_p2s = sorted(ratings, key=lambda p2_tup: abs(p1[1] - p2_tup[1]))[1:20]
+        choices = random.sample(candidate_p2s, per_n)
+        print("Pairing {}, sigma {:.2f}".format(p1[0], p1[2]))
+        for p2 in choices:
+            res.append([p1[0], p2[0]])
+            print("   {}, ratings delta {:.2f}".format(p2[0], abs(p1[1]-p2[1])))
+    return res
+
+
+def sync(root, force_all=False):
+    import subprocess
+    last_ts = last_timestamp()
+    if last_ts and not force_all:
+        # Build a list of days from the day before our last timestamp to today
+        num_days = (dt.datetime.utcnow() -
+                    dt.datetime.utcfromtimestamp(last_ts) +
+                    dt.timedelta(days=1)).days
+        ds = [(dt.datetime.utcnow() - dt.timedelta(days=d)).strftime("%Y-%m-%d") for d in range(num_days+1)]
+        for d in ds:
+            if not os.path.isdir(os.path.join(root, d)):
+                os.mkdir(os.path.join(root, d))
+            cmd = ["gsutil", "-m", "rsync", "-r", os.path.join(fsdb.eval_dir(), d), os.path.join(root, d)]
+            print(" ".join(cmd))
+            subprocess.call(cmd)
+
+        ingest_dirs(root, ds)
+    else:
+        cmd = "gsutil -m rsync -r {0} {1}".format(fsdb.eval_dir(), root)
+        print(cmd)
+        subprocess.call(cmd.split())
+        dirs = os.listdir(root)
+        ingest_dirs(root, dirs)
+
+
+def main():
+    root = os.path.abspath(os.path.join("sgf", fsdb.FLAGS.bucket_name, "sgf/eval"))
+    sync(root, True)
+    models = fsdb.get_models()
+    r = compute_ratings()
+    for v, k in sorted([(v, k) for k, v in r.items()])[-20:][::-1]:
+        print(models[model_num_for(k)][1], v)
+    db = sqlite3.connect("ratings.db")
+    print(db.execute("select count(*) from wins").fetchone()[0], "games")
+    for m in models[-10:]:
+        try:
+            print(m[1], r[model_id(m[0])])
+        except KeyError:
```

Comment:
m_id = model_id(m[0])
print(m[1], r.get(m_id, "model id not found({})".format(m_id))

then you can drop the try except